Runtime/Scripts/DataCollection/RGDataCollection.cs

```diff
@@ -332,18 +383,47 @@ public void Cleanup()
          */
         private void Cleanup(long clientId)
         {
-            _clientIdToValidationSummaries.TryRemove(clientId, out _);
-            _clientIdToValidationDataTask.TryRemove(clientId, out _);
-            _clientIdToReplayDataTasks.TryRemove(clientId, out _);
-            _clientIdToBots.Remove(clientId, out _);
-            _clientIdStartTimes.Remove(clientId, out _);
             var replayPath = GetSessionDirectory($"replayData/{clientId}");
             Directory.Delete(replayPath, true);
             RGDebug.LogVerbose($"DataCollection[{clientId}] - Cleaned up all data for client");
         }
 
+        private static string SerializeJsonLines(IEnumerable<object> items)
+        {
+            // Convert the list into JSONL format
+            var builder = new StringBuilder();
+            foreach (var item in items)
+            {
+                builder.AppendLine(JsonConvert.SerializeObject(item));
+            }
+
+            // Combine the JSON strings with newline characters
+            return builder.ToString();
+        }
+
+        private void LogMessageReceived(string message, string stacktrace, LogType type)
+        {
+            if (_clientIdToState.IsEmpty)
+            {
+                // Avoid allocating a LogDataPoint if nobody is listening.
+                return;
+            }
+
+            var dataPoint = new LogDataPoint(DateTimeOffset.Now, type, message, stacktrace);
+
+            // Note: Enumerating a concurrent dictionary is _safe_ in that it won't fail,
+            // but it doesn't take a snapshot of the dictionary, so entries may be added/removed _during_ enumeration.
+            // That should be fine for us.
+            // The main edge case here is a bot stopping _during_ this enumeration,
+            // which would cause us to add a log entry to the queue after it's been flushed to the server.
+            // That's fine, it just means we'll miss that log entry. The queue will still get garbage collected.
+            foreach (var (_, state) in _clientIdToState)
+            {
+                state.logs.Enqueue(dataPoint);
```

Comment:
Can you do some testing in the unity profiler and check how many ms this is taking.

When micro-optimizing the Platformer bot libraries I found that in cases like this is was actually more efficient to lock on the clientId specific object outside the loop and use a non concurrent collection type.  The overhead of the concurrent collection for all the adds took much more frametime than the lock operation.

Where I saw that the worst was on ConcurrentDictionary, but that's why I ask if you could check this in the profiler as even a single ms reduction in frame time is huge.  Our SDK should strive to take as close to zero time as possible so that the game developers don't have to worry about framerate because of us.

Reply:
Yeah, that's a good point. ConcurrentDictionary is actually quite a lock-heavy data structure. I'll take a look in the profiler. I also have an idea for how to remove the dictionary in this case, by having each bot set up it's own handler on `logMessageReceivedThreaded` (since events can have multiple handlers, which can be added/removed at will) and have that handler capture the "state" object in a closure.